### PR TITLE
re additions for buffer (cstring) RE matching

### PR DESCRIPTION
**Pull request by jlp765**
These changes are to enable RE procs to work on cstring/array[..., char] without converting to strings first.

This also makes it easier to use RE's with memfile char data

### CODE REVIEW

**Comment by Araq** [comment]
Patch is ok, but if you introduce 'cstring' variants, please make the string variants call the cstring variants in order to fight code size.

**Comment by jlp765** [comment]
>make the string variants call the cstring variants in order to fight code size.

?? can you give an example

**Comment by Araq** [comment]
```nim

proc aa(x: cstring) = ... impl here...
proc aa(x: string) = aa(cstring(x)) # call cstring variant
```

**Comment by jlp765** [comment]
Ok, I comprehend now.

The issue is that I used cstring as a way to refer to a char buffer.

**Requirement**:
To work with memfile MemSlices and similar char buffers (that are not necessarily null-terminated strings), but "arrays" of char that have a size,   
ie search through this char buffer containing "size" chars 
(and even if the buffer contains cstrings, search through a defined number of chars independent of c-style strings)

Should I have used pointer rather than cstring?

**Comment by Araq** [comment]
I don't see the problem with ``cstring``, the same applies:

```nim
proc aa(x: cstring; L: int) = ... impl here...
proc aa(x: string) = aa(cstring(x), x.len) # call cstring variant

```

File: lib/impure/re.nim [hunk 1]
```diff
@@ -150,80 +170,165 @@ proc findBounds*(s: string, pattern: Regex,
...
     rawMatches = rtarray.getRawData
-    res = pcre.exec(pattern.h, pattern.e, s, len(s).cint, start.cint, 0'i32,
+    res = pcre.exec(pattern.h, pattern.e, buf, bufSize.cint, start.cint, 0'i32,
       cast[ptr cint](rawMatches), (matches.len+1).cint*3)
   if res < 0'i32: return res
   for i in 1..int(res)-1:
     var a = rawMatches[i * 2]
     var b = rawMatches[i * 2 + 1]
-    if a >= 0'i32: matches[i-1] = substr(s, int(a), int(b)-1)
+    if a >= 0'i32: matches[i-1] = cast[string](buf).substr(int(a), int(b)-1)
```
Comment (Araq) [inline lib/impure/re.nim]: You cannot use 'cast' here!

File: lib/impure/re.nim [hunk 2]
```diff
@@ -243,15 +348,36 @@ iterator findAll*(s: string, pattern: Regex, start = 0): string =
     yield substr(s, int(a), int(b)-1)
     i = b
 
+iterator findAll*(buf: var cstring, pattern: Regex, start = 0, bufSize: int): string =
```
Comment (Araq) [inline lib/impure/re.nim]: buf: var cstring? Why?

**Comment by Araq** [comment]
Ping?

**Pull request merged**